diff --git a/app/welcomepage/Heading.qml b/app/welcomepage/Heading.qml
index c38fd4f..5c9e1cd 100644
--- a/app/welcomepage/Heading.qml
+++ b/app/welcomepage/Heading.qml
@@ -1,5 +1,25 @@
+/* KDevelop
+ *
+ * Copyright 2011 Aleix Pol <aleixpol@kde.org>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
 import org.kde.plasma.components 0.1
 
 Label {
-    font.pointSize: theme.defaultFont.pointSize*Math.log(theme.defaultFont.pointSize)
-}
\ No newline at end of file
+    font.pointSize: theme.defaultFont.pointSize*1.5
+}
diff --git a/app/welcomepage/Link.qml b/app/welcomepage/Link.qml
index b626447..5a8a931 100644
--- a/app/welcomepage/Link.qml
+++ b/app/welcomepage/Link.qml
@@ -22,4 +22,4 @@ import QtQuick 1.1
 import org.kde.plasma.components 0.1
 
 ToolButton {
-}
\ No newline at end of file
+}
diff --git a/app/welcomepage/StandardBackground.qml b/app/welcomepage/StandardBackground.qml
index 503d875..8a60d8e 100644
--- a/app/welcomepage/StandardBackground.qml
+++ b/app/welcomepage/StandardBackground.qml
@@ -38,14 +38,19 @@ Rectangle
             left: parent.left
             margins: parent.margins
         }
-        width: 256
-        height: 100
+        width: toolsLoader.width + 2*toolsLoader.margins
+        height: toolsLoader.height + 2*toolsLoader.margins
         Loader {
-            sourceComponent: tools
+            id: toolsLoader
+
+            property int margins: 20
+
             anchors {
-                fill: parent
-                margins: 20
+                top: parent.top
+                left: parent.left
+                margins: toolsLoader.margins
             }
+            sourceComponent: tools
         }
     }
     
diff --git a/app/welcomepage/code.qml b/app/welcomepage/code.qml
index 5cb4649..832bcad 100644
--- a/app/welcomepage/code.qml
+++ b/app/welcomepage/code.qml
@@ -32,17 +32,14 @@ StandardBackground {
 
     tools: ButtonColumn {
         spacing: 10
-        Link {
-            anchors.margins: 20
-            width: parent.width
 
+        Link {
+            iconSource: "applications-development"
             text: i18n("Starting")
             onClicked: root.state = "starting"
         }
         Link {
-            anchors.margins: 20
-            width: parent.width
-
+            iconSource: "help-contents"
             text: i18n("Learn KDevelop!")
             onClicked: root.state = "learning"
         }
diff --git a/debuggers/gdb/breakpointcontroller.cpp b/debuggers/gdb/breakpointcontroller.cpp
index b974595..4a41c36 100644
--- a/debuggers/gdb/breakpointcontroller.cpp
+++ b/debuggers/gdb/breakpointcontroller.cpp
@@ -230,17 +230,26 @@ void BreakpointController::handleBreakpointListInitial(const GDBMI::ResultRecord
                 if (condition != b->condition())
                     continue;
 
-                QString location = mi_b["original-location"].literal();
-                kDebug() << "location" << location;
-                QRegExp rx("^(.+):(\\d+)$");
-                if (rx.indexIn(location) != -1) {
-                    if (unquoteExpression(rx.cap(1)) == b->url().pathOrUrl(KUrl::RemoveTrailingSlash) && rx.cap(2).toInt()-1 == b->line()) {
+                QString location;
+                if (mi_b.hasField("original-location")) {
+                    location = mi_b["original-location"].literal();
+                    kDebug() << "location" << location;
+                    QRegExp rx("^(.+):(\\d+)$");
+                    if (rx.indexIn(location) != -1) {
+                        if (unquoteExpression(rx.cap(1)) == b->url().pathOrUrl(KUrl::RemoveTrailingSlash) && rx.cap(2).toInt() - 1 == b->line()) {
+                            updateBreakpoint = b;
+                        } else {
+                            kDebug() << "!=" << b->location();
+                        }
+                    } else if (location == b->location()) {
                         updateBreakpoint = b;
-                    } else {
-                        kDebug() << "!=" << b->location();
                     }
-                } else if (location == b->location()) {
-                    updateBreakpoint = b;
+                }else if (mi_b.hasField("what") && mi_b["what"].literal() == "exception throw") {
+                    if (b->expression() == "catch throw") {
+                        updateBreakpoint = b;
+                    }
+                }else{
+                    kWarning() << "That's too bad, breakpoint doesn't contain \"original-location\" field ";
                 }
             }
             if (updateBreakpoint) break;
@@ -309,10 +318,17 @@ void BreakpointController::sendMaybe(KDevelop::Breakpoint* breakpoint)
                 } else {
                     location = breakpoint->location();
                 }
-                debugSession()->addCommandToFront(
+                if (breakpoint->expression() == "catch throw") {
+                    debugSession()->addCommand(
+                    new GDBCommand(GDBMI::NonMI,
+                                location));
+                    breakpoint->setDeleted();
+                }else{
+                   debugSession()->addCommandToFront(
                     new GDBCommand(BreakInsert,
                                 quoteExpression(location),
                                 new InsertedHandler(this, breakpoint)));
+                }
                 addedCommand = true;
             } else {
                 QString opt;
@@ -460,11 +476,12 @@ void BreakpointController::update(KDevelop::Breakpoint *breakpoint, const GDBMI:
                 breakpoint->setData(KDevelop::Breakpoint::LocationColumn, unquoteExpression(location));
             }
         }
+    } else if (b.hasField("what") && b["what"].literal() == "exception throw") {
+        breakpoint->setExpression("catch throw");
     } else {
-        breakpoint->setData(KDevelop::Breakpoint::LocationColumn, "Your GDB is too old");
+        kWarning() << "That's too bad, breakpoint doesn't contain \"original-location\" field ";
     }
 
-
     if (!m_dirty[breakpoint].contains(KDevelop::Breakpoint::ConditionColumn)
         && !m_errors[breakpoint].contains(KDevelop::Breakpoint::ConditionColumn))
     {
diff --git a/debuggers/gdb/debugsession.cpp b/debuggers/gdb/debugsession.cpp
index f1c0145..5207c25 100644
--- a/debuggers/gdb/debugsession.cpp
+++ b/debuggers/gdb/debugsession.cpp
@@ -1426,17 +1426,11 @@ void DebugSession::programRunning()
 
 void DebugSession::handleVersion(const QStringList& s)
 {
-    const int minVersion1 = 7;
-    const int minVersion2 = 0;
-    const int minVersion3 = 0;
-
     kDebug() << s.first();
-    QRegExp rx("([0-9]+)\\.([0-9]+)(?:\\.([0-9]+))?");
-    rx.indexIn(s.first());
-    if (rx.cap(1).toInt() < minVersion1
-        || (rx.cap(1).toInt() == minVersion1 && (rx.cap(2).toInt() < minVersion2
-            || ( rx.cap(2).toInt() == minVersion2 && minVersion3 > 0 && rx.numCaptures() == 3
-                && rx.cap(3).toInt() < minVersion3))))
+    // minimal version is 7.0,0
+    QRegExp rx("([7-9]+)\\.([0-9]+)(\\.([0-9]+))?");
+    int idx = rx.indexIn(s.first());
+    if (idx == -1)
     {
         if (qApp->type() == QApplication::Tty)  {
             //for unittest
diff --git a/debuggers/gdb/gdbvariable.cpp b/debuggers/gdb/gdbvariable.cpp
index a98e25d..dae1e89 100644
--- a/debuggers/gdb/gdbvariable.cpp
+++ b/debuggers/gdb/gdbvariable.cpp
@@ -206,7 +206,8 @@ public:
                     GdbVariable* var = static_cast<GdbVariable*>(xvar);
                     var->setTopLevel(false);
                     var->setVarobj(child["name"].literal());
-                    var->setHasMoreInitial(child["numchild"].toInt());
+                    bool hasMore = child["numchild"].toInt() != 0 || ( child.hasField("dynamic") && child["dynamic"].toInt()!=0 );
+                    var->setHasMoreInitial(hasMore);
                     variable->appendChild(var);
                     var->setValue(child["value"].literal());
                 }
@@ -306,7 +307,8 @@ void GdbVariable::handleUpdate(const GDBMI::Value& var)
                 GdbVariable* var = static_cast<GdbVariable*>(xvar);
                 var->setTopLevel(false);
                 var->setVarobj(child["name"].literal());
-                var->setHasMoreInitial(child["numchild"].toInt() != 0);
+                bool hasMore = child["numchild"].toInt() != 0 || ( child.hasField("dynamic") && child["dynamic"].toInt()!=0 );
+                var->setHasMoreInitial(hasMore);
                 appendChild(var);
                 var->setValue(child["value"].literal());
                 var->setChanged(true);
diff --git a/documentation/CMakeLists.txt b/documentation/CMakeLists.txt
index 5227063..66f1c5c 100644
--- a/documentation/CMakeLists.txt
+++ b/documentation/CMakeLists.txt
@@ -1,2 +1,6 @@
-macro_optional_add_subdirectory(qthelp)
+macro_log_feature(QT_QTHELP_FOUND "QtHelp" "The Help module for the Qt toolkit, needed for the qthelp plugin" "http://qt-project.org/" FALSE)
+
+if(QT_QTHELP_FOUND)
+ macro_optional_add_subdirectory(qthelp)
+endif()
 macro_optional_add_subdirectory(manpage)
diff --git a/documentation/qthelp/qthelpconfig.cpp b/documentation/qthelp/qthelpconfig.cpp
index 6ef58d5..8ee5271 100644
--- a/documentation/qthelp/qthelpconfig.cpp
+++ b/documentation/qthelp/qthelpconfig.cpp
@@ -317,6 +317,9 @@ void QtHelpConfig::down()
 
 void QtHelpConfig::knsUpdate(KNS3::Entry::List list)
 {
+    if (list.isEmpty())
+        return;
+
     foreach (const KNS3::Entry& e, list) {
         if(e.status() == KNS3::Entry::Installed) {
             if(e.installedFiles().size() == 1) {
diff --git a/file_templates/classes/c_gobject/c_gobject.desktop b/file_templates/classes/c_gobject/c_gobject.desktop
index 4bfe617..2d12b34 100644
--- a/file_templates/classes/c_gobject/c_gobject.desktop
+++ b/file_templates/classes/c_gobject/c_gobject.desktop
@@ -16,6 +16,7 @@ Name[hu]=GObject
 Name[it]=GObject
 Name[kk]=GObject
 Name[nb]=GObjekt
+Name[nds]=GObject
 Name[nl]=GObject
 Name[pl]=GObject
 Name[pt]=GObject
@@ -77,6 +78,7 @@ Language[it]=C
 Language[kk]=C
 Language[mr]=C
 Language[nb]=C
+Language[nds]=C
 Language[nl]=C
 Language[nn]=C
 Language[pl]=C
@@ -146,6 +148,7 @@ Name[hu]=Megvalósítás
 Name[it]=Implementazione
 Name[kk]=Іске асыруы
 Name[nb]=Implementering
+Name[nds]=Ümsetten
 Name[nl]=Implementatie
 Name[pl]=Implementacja
 Name[pt]=Implementação
diff --git a/file_templates/classes/c_gobject_private/c_gobject_private.desktop b/file_templates/classes/c_gobject_private/c_gobject_private.desktop
index 4e9ebc4..8a9f081 100644
--- a/file_templates/classes/c_gobject_private/c_gobject_private.desktop
+++ b/file_templates/classes/c_gobject_private/c_gobject_private.desktop
@@ -76,6 +76,7 @@ Language[it]=C
 Language[kk]=C
 Language[mr]=C
 Language[nb]=C
+Language[nds]=C
 Language[nl]=C
 Language[nn]=C
 Language[pl]=C
@@ -124,9 +125,6 @@ Name[uk]=Заголовок
 Name[x-test]=xxHeaderxx
 Name[zh_CN]=头文件
 Name[zh_TW]=標頭
-<<<<<<< HEAD
-=======
->>>>>>> 4.5
 File=class.h
 OutputFile={% for ns in namespaces %}{{ ns }}_{% endfor %}{{ name }}.h
 
@@ -148,6 +146,7 @@ Name[hu]=Megvalósítás
 Name[it]=Implementazione
 Name[kk]=Іске асыруы
 Name[nb]=Implementering
+Name[nds]=Ümsetten
 Name[nl]=Implementatie
 Name[pl]=Implementacja
 Name[pt]=Implementação
diff --git a/file_templates/classes/c_gobject_properties/c_gobject_properties.desktop b/file_templates/classes/c_gobject_properties/c_gobject_properties.desktop
index 356664c07..c24db16 100644
--- a/file_templates/classes/c_gobject_properties/c_gobject_properties.desktop
+++ b/file_templates/classes/c_gobject_properties/c_gobject_properties.desktop
@@ -76,6 +76,7 @@ Language[it]=C
 Language[kk]=C
 Language[mr]=C
 Language[nb]=C
+Language[nds]=C
 Language[nl]=C
 Language[nn]=C
 Language[pl]=C
@@ -145,6 +146,7 @@ Name[hu]=Megvalósítás
 Name[it]=Implementazione
 Name[kk]=Іске асыруы
 Name[nb]=Implementering
+Name[nds]=Ümsetten
 Name[nl]=Implementatie
 Name[pl]=Implementacja
 Name[pt]=Implementação
diff --git a/file_templates/classes/cpp_basic/cpp_basic.desktop b/file_templates/classes/cpp_basic/cpp_basic.desktop
index 0987565..c17b613 100644
--- a/file_templates/classes/cpp_basic/cpp_basic.desktop
+++ b/file_templates/classes/cpp_basic/cpp_basic.desktop
@@ -79,6 +79,7 @@ Language[it]=C++
 Language[kk]=C++
 Language[mr]=C++
 Language[nb]=C++
+Language[nds]=C++
 Language[nl]=C++
 Language[nn]=C++
 Language[pl]=C++
@@ -128,9 +129,6 @@ Name[uk]=Заголовок
 Name[x-test]=xxHeaderxx
 Name[zh_CN]=头文件
 Name[zh_TW]=標頭
-<<<<<<< HEAD
-=======
->>>>>>> 4.5
 File=class.h
 OutputFile={{ name }}.h
 
@@ -152,6 +150,7 @@ Name[hu]=Megvalósítás
 Name[it]=Implementazione
 Name[kk]=Іске асыруы
 Name[nb]=Implementering
+Name[nds]=Ümsetten
 Name[nl]=Implementatie
 Name[pl]=Implementacja
 Name[pt]=Implementação
diff --git a/file_templates/classes/private_pointer/private_pointer.desktop b/file_templates/classes/private_pointer/private_pointer.desktop
index 11edd7c..4307978 100644
--- a/file_templates/classes/private_pointer/private_pointer.desktop
+++ b/file_templates/classes/private_pointer/private_pointer.desktop
@@ -75,6 +75,7 @@ Language[it]=C++
 Language[kk]=C++
 Language[mr]=C++
 Language[nb]=C++
+Language[nds]=C++
 Language[nl]=C++
 Language[nn]=C++
 Language[pl]=C++
@@ -156,9 +157,6 @@ Name[uk]=Закритий (private) заголовок
 Name[x-test]=xxPrivate Headerxx
 Name[zh_CN]=私有头文件
 Name[zh_TW]=私密標頭
-<<<<<<< HEAD
-=======
->>>>>>> 4.5
 File=class_p.h
 OutputFile={{ name }}_p.h
 
@@ -180,6 +178,7 @@ Name[hu]=Megvalósítás
 Name[it]=Implementazione
 Name[kk]=Іске асыруы
 Name[nb]=Implementering
+Name[nds]=Ümsetten
 Name[nl]=Implementatie
 Name[pl]=Implementacja
 Name[pt]=Implementação
diff --git a/file_templates/classes/python_basic/python_basic.desktop b/file_templates/classes/python_basic/python_basic.desktop
index a3ef294..0ff4f87 100644
--- a/file_templates/classes/python_basic/python_basic.desktop
+++ b/file_templates/classes/python_basic/python_basic.desktop
@@ -78,6 +78,7 @@ Language[it]=Python
 Language[kk]=Python
 Language[mr]=पायथोन
 Language[nb]=Python
+Language[nds]=Python
 Language[nl]=Python
 Language[nn]=Python
 Language[pl]=Python
@@ -113,6 +114,7 @@ Name[hu]=Definíció
 Name[it]=Definizione
 Name[kk]=Анықтама
 Name[nb]=Definisjon
+Name[nds]=Definitschoon
 Name[nl]=Definitie
 Name[pl]=Definicja
 Name[pt]=Definição
diff --git a/file_templates/classes/qobject/qobject.desktop b/file_templates/classes/qobject/qobject.desktop
index d5e1db6..67183e4 100644
--- a/file_templates/classes/qobject/qobject.desktop
+++ b/file_templates/classes/qobject/qobject.desktop
@@ -15,6 +15,7 @@ Name[hu]=QObject alosztály
 Name[it]=Sottoclasse GObject
 Name[kk]=QObject субкласы
 Name[nb]=QObject subklasse
+Name[nds]=QObject-Ünnerklass
 Name[nl]=Subklasse van QObject
 Name[pl]=Podklasa QObject
 Name[pt]=Sub-classe de QObject
@@ -43,6 +44,7 @@ Comment[hu]=QObject alosztály tulajdonságokkal
 Comment[it]=Sottoclasse GObject con delle proprietà
 Comment[kk]=Қасиеттері бар QObject субкласы
 Comment[nb]=QObject subklasse med egenskaper
+Comment[nds]=QObject-Ünnerklass mit Egenschappen
 Comment[nl]=Subklasse van QObject met eigenschappen
 Comment[pl]=Podklasa QObject z właściwościami
 Comment[pt]=Sub-classe de QObject com propriedades
@@ -74,6 +76,7 @@ Language[it]=C++
 Language[kk]=C++
 Language[mr]=C++
 Language[nb]=C++
+Language[nds]=C++
 Language[nl]=C++
 Language[nn]=C++
 Language[pl]=C++
@@ -144,6 +147,7 @@ Name[hu]=Megvalósítás
 Name[it]=Implementazione
 Name[kk]=Іске асыруы
 Name[nb]=Implementering
+Name[nds]=Ümsetten
 Name[nl]=Implementatie
 Name[pl]=Implementacja
 Name[pt]=Implementação
diff --git a/file_templates/classes/qt_shared/qt_shared.desktop b/file_templates/classes/qt_shared/qt_shared.desktop
index 07cc9f0..a81f993 100644
--- a/file_templates/classes/qt_shared/qt_shared.desktop
+++ b/file_templates/classes/qt_shared/qt_shared.desktop
@@ -74,6 +74,7 @@ Language[it]=C++
 Language[kk]=C++
 Language[mr]=C++
 Language[nb]=C++
+Language[nds]=C++
 Language[nl]=C++
 Language[nn]=C++
 Language[pl]=C++
@@ -143,6 +144,7 @@ Name[hu]=Megvalósítás
 Name[it]=Implementazione
 Name[kk]=Іске асыруы
 Name[nb]=Implementering
+Name[nds]=Ümsetten
 Name[nl]=Implementatie
 Name[pl]=Implementacja
 Name[pt]=Implementação
diff --git a/file_templates/classes/qt_widget/qt_widget.desktop b/file_templates/classes/qt_widget/qt_widget.desktop
index abd5088..51dd886 100644
--- a/file_templates/classes/qt_widget/qt_widget.desktop
+++ b/file_templates/classes/qt_widget/qt_widget.desktop
@@ -74,6 +74,7 @@ Language[it]=C++
 Language[kk]=C++
 Language[mr]=C++
 Language[nb]=C++
+Language[nds]=C++
 Language[nl]=C++
 Language[nn]=C++
 Language[pl]=C++
@@ -144,6 +145,7 @@ Name[hu]=Megvalósítás
 Name[it]=Implementazione
 Name[kk]=Іске асыруы
 Name[nb]=Implementering
+Name[nds]=Ümsetten
 Name[nl]=Implementatie
 Name[pl]=Implementacja
 Name[pt]=Implementação
@@ -178,6 +180,7 @@ Name[it]=Interfaccia utente
 Name[kk]=Пайдаланушы интерфейсі
 Name[mr]=वापरकर्ता संवाद
 Name[nb]=Brukerflate
+Name[nds]=Böversiet
 Name[nl]=Gebruikersinterface
 Name[pl]=Interfejs użytkownika
 Name[pt]=Interface do Utilizador
diff --git a/file_templates/other/cmake_module/cmake_module.desktop b/file_templates/other/cmake_module/cmake_module.desktop
index 87fa169..4150b42 100644
--- a/file_templates/other/cmake_module/cmake_module.desktop
+++ b/file_templates/other/cmake_module/cmake_module.desktop
@@ -73,6 +73,7 @@ Language[hu]=CMake
 Language[it]=CMake
 Language[kk]=CMake
 Language[nb]=CMake
+Language[nds]=CMake
 Language[nl]=CMake
 Language[nn]=CMake
 Language[pl]=CMake
@@ -109,6 +110,7 @@ Name[it]=Modulo
 Name[kk]=Модулі
 Name[mr]=विभाग
 Name[nb]=Modul
+Name[nds]=Moduul
 Name[nl]=Module
 Name[pl]=Moduł
 Name[pt]=Módulo
diff --git a/file_templates/testing/cpp_qtestlib/cpp_qtestlib.desktop b/file_templates/testing/cpp_qtestlib/cpp_qtestlib.desktop
index 64b499e..5f01fc9 100644
--- a/file_templates/testing/cpp_qtestlib/cpp_qtestlib.desktop
+++ b/file_templates/testing/cpp_qtestlib/cpp_qtestlib.desktop
@@ -15,8 +15,9 @@ Name[hu]=Teszt alkalmazáscsomag
 Name[it]=Gruppi di test
 Name[kk]=Сынақ дестесі
 Name[nb]=Testsuite
+Name[nds]=Testümgeven
 Name[nl]=Testsuite
-Name[pl]=Zestaw testowy
+Name[pl]=Zestaw próbny
 Name[pt]=Conjunto de Testes
 Name[pt_BR]=Conjunto de testes
 Name[ru]=Комплект тестов
@@ -74,6 +75,7 @@ Language[it]=C++
 Language[kk]=C++
 Language[mr]=C++
 Language[nb]=C++
+Language[nds]=C++
 Language[nl]=C++
 Language[nn]=C++
 Language[pl]=C++
@@ -144,6 +146,7 @@ Name[hu]=Megvalósítás
 Name[it]=Implementazione
 Name[kk]=Іске асыруы
 Name[nb]=Implementering
+Name[nds]=Ümsetten
 Name[nl]=Implementatie
 Name[pl]=Implementacja
 Name[pt]=Implementação
diff --git a/file_templates/testing/cpp_qtestlib_kde/cpp_qtestlib_kde.desktop b/file_templates/testing/cpp_qtestlib_kde/cpp_qtestlib_kde.desktop
index 5f63810..a4e3de7 100644
--- a/file_templates/testing/cpp_qtestlib_kde/cpp_qtestlib_kde.desktop
+++ b/file_templates/testing/cpp_qtestlib_kde/cpp_qtestlib_kde.desktop
@@ -15,6 +15,7 @@ Name[hu]=QTestLib KDE-vel
 Name[it]=QTestLib con KDE
 Name[kk]=KDE QTestLib
 Name[nb]=QTestLib med KDE
+Name[nds]=QTestLib mit KDE
 Name[nl]=QTestLib met KDE
 Name[pl]=QTestLib z KDE
 Name[pt]=QTestLib com o KDE
@@ -74,6 +75,7 @@ Language[it]=C++
 Language[kk]=C++
 Language[mr]=C++
 Language[nb]=C++
+Language[nds]=C++
 Language[nl]=C++
 Language[nn]=C++
 Language[pl]=C++
@@ -144,6 +146,7 @@ Name[hu]=Megvalósítás
 Name[it]=Implementazione
 Name[kk]=Іске асыруы
 Name[nb]=Implementering
+Name[nds]=Ümsetten
 Name[nl]=Implementatie
 Name[pl]=Implementacja
 Name[pt]=Implementação
diff --git a/file_templates/testing/cpp_qtestlib_kdevelop/cpp_qtestlib_kdevelop.desktop b/file_templates/testing/cpp_qtestlib_kdevelop/cpp_qtestlib_kdevelop.desktop
index 318c8f6..964d4c4 100644
--- a/file_templates/testing/cpp_qtestlib_kdevelop/cpp_qtestlib_kdevelop.desktop
+++ b/file_templates/testing/cpp_qtestlib_kdevelop/cpp_qtestlib_kdevelop.desktop
@@ -74,6 +74,7 @@ Language[it]=C++
 Language[kk]=C++
 Language[mr]=C++
 Language[nb]=C++
+Language[nds]=C++
 Language[nl]=C++
 Language[nn]=C++
 Language[pl]=C++
@@ -144,6 +145,7 @@ Name[hu]=Megvalósítás
 Name[it]=Implementazione
 Name[kk]=Іске асыруы
 Name[nb]=Implementering
+Name[nds]=Ümsetten
 Name[nl]=Implementatie
 Name[pl]=Implementacja
 Name[pt]=Implementação
diff --git a/file_templates/testing/php_phpunit/php_phpunit.desktop b/file_templates/testing/php_phpunit/php_phpunit.desktop
index ef674a3..891a437 100644
--- a/file_templates/testing/php_phpunit/php_phpunit.desktop
+++ b/file_templates/testing/php_phpunit/php_phpunit.desktop
@@ -74,6 +74,7 @@ Language[hu]=PHP
 Language[it]=Php
 Language[kk]=Php
 Language[nb]=Php
+Language[nds]=PHP
 Language[nl]=PHP
 Language[pl]=Php
 Language[pt]=PHP
@@ -87,9 +88,6 @@ Language[uk]=PHP
 Language[x-test]=xxPhpxx
 Language[zh_CN]=Php
 Language[zh_TW]=Php
-<<<<<<< HEAD
-=======
->>>>>>> 4.5
 Type=Test
 Files=Implementation
 
@@ -111,6 +109,7 @@ Name[hu]=Megvalósítás
 Name[it]=Implementazione
 Name[kk]=Іске асыруы
 Name[nb]=Implementering
+Name[nds]=Ümsetten
 Name[nl]=Implementatie
 Name[pl]=Implementacja
 Name[pt]=Implementação
diff --git a/file_templates/testing/python_pyunit/python_pyunit.desktop b/file_templates/testing/python_pyunit/python_pyunit.desktop
index 2380644..6b44b6d 100644
--- a/file_templates/testing/python_pyunit/python_pyunit.desktop
+++ b/file_templates/testing/python_pyunit/python_pyunit.desktop
@@ -16,6 +16,7 @@ Name[hu]=PyUnit
 Name[it]=PyUnit
 Name[kk]=PyUnit
 Name[nb]=PyUnit
+Name[nds]=PyUnit
 Name[nl]=PyUnit
 Name[pl]=PyUnit
 Name[pt]=PyUnit
@@ -76,6 +77,7 @@ Language[it]=Python
 Language[kk]=Python
 Language[mr]=पायथोन
 Language[nb]=Python
+Language[nds]=Python
 Language[nl]=Python
 Language[nn]=Python
 Language[pl]=Python
@@ -111,6 +113,7 @@ Name[hu]=Megvalósítás
 Name[it]=Implementazione
 Name[kk]=Іске асыруы
 Name[nb]=Implementering
+Name[nds]=Ümsetten
 Name[nl]=Implementatie
 Name[pl]=Implementacja
 Name[pt]=Implementação
diff --git a/kdeintegration/executeplasmoid/kdevexecuteplasmoid.desktop.cmake b/kdeintegration/executeplasmoid/kdevexecuteplasmoid.desktop.cmake
index dd73d82..df2595e 100644
--- a/kdeintegration/executeplasmoid/kdevexecuteplasmoid.desktop.cmake
+++ b/kdeintegration/executeplasmoid/kdevexecuteplasmoid.desktop.cmake
@@ -1,5 +1,6 @@
 [Desktop Entry]
 Type=Service
+Icon=system-run
 Exec=blubb
 Comment=Lets you view a plasmoid
 Comment[bs]=Omogućava da vidite plazmoid
@@ -45,6 +46,7 @@ Name[hu]=Plasmoid indító
 Name[it]=Lanciatore di plasmoidi
 Name[kk]=Плазмоид жеккіші
 Name[nb]=Plasmoidestarter
+Name[nds]=Oproper för Plasma-Lüttprogrammen
 Name[nl]=Programmastarter van plasmoid
 Name[pl]=Program uruchamiający plazmoid
 Name[pt]=Lançamento de Plasmóides
diff --git a/languages/cpp/codecompletion/context.cpp b/languages/cpp/codecompletion/context.cpp
index d851050..99c91cf 100644
--- a/languages/cpp/codecompletion/context.cpp
+++ b/languages/cpp/codecompletion/context.cpp
@@ -1302,7 +1302,7 @@ void getOverridable(DUContext* base, DUContext* current, QMap< QPair<IndexedType
   foreach(Declaration* decl, current->localDeclarations()) {
     ClassFunctionDeclaration* classFun = dynamic_cast<ClassFunctionDeclaration*>(decl);
     // one can only override the direct parent's ctor
-    if(classFun && (classFun->isVirtual() || (depth == 0 && classFun->isConstructor()))) {
+    if(classFun && (classFun->isVirtual() || (depth == 0 && classFun->isConstructor())) && !classFun->isExplicitlyDeleted()) {
       QPair<IndexedType, IndexedString> key = qMakePair(classFun->indexedType(), classFun->identifier().identifier());
       if(base->owner()) {
         if(classFun->isConstructor() || classFun->isDestructor())
@@ -1355,7 +1355,7 @@ QList<CompletionTreeItemPointer> CodeCompletionContext::memberAccessCompletionIt
     ifDebug( kDebug() << "container:" << ctx->scopeIdentifier(true).toString(); )
 
     QList<DeclarationDepthPair> decls = ctx->allDeclarations(ctx->range().end, m_duContext->topContext(), false );
-    decls += namespaceItems(ctx, ctx->range().end, false);
+    decls += namespaceItems(ctx, ctx->range().end, false, containers);
 
     foreach( const DeclarationDepthPair& decl, Cpp::hideOverloadedDeclarations(decls, typeIsConst ) )
     {
@@ -1813,7 +1813,8 @@ QList<DeclAccessPair> CodeCompletionContext::getLookaheadMatches(Declaration* fo
   return ret;
 }
 
-QList<DeclarationDepthPair> CodeCompletionContext::namespaceItems(DUContext* duContext, const CursorInRevision& position, bool global) const
+QList<DeclarationDepthPair> CodeCompletionContext::namespaceItems(DUContext* duContext, const CursorInRevision& position,
+                                                                  bool global, const QSet<DUContext*>& skipContexts) const
 {
   QList<DeclarationDepthPair> decls;
   QList<Declaration*> foundDecls;
@@ -1857,6 +1858,9 @@ QList<DeclarationDepthPair> CodeCompletionContext::namespaceItems(DUContext* duC
       if(contextDecl->kind() != Declaration::Namespace || !contextDecl->internalContext())
         continue;
       DUContext* context = contextDecl->internalContext();
+      if (skipContexts.contains(context)) {
+        continue;
+      }
 
       if(context->range().contains(duContext->range()) && context->url() == duContext->url())
         continue; //If the context surrounds the current one, the declarations are visible through allDeclarations(..).
diff --git a/languages/cpp/codecompletion/context.h b/languages/cpp/codecompletion/context.h
index 8e0ca69..3981b91 100644
--- a/languages/cpp/codecompletion/context.h
+++ b/languages/cpp/codecompletion/context.h
@@ -314,7 +314,8 @@ namespace Cpp {
       ///Handle code-completion for constructor-initializers, \returns true if valid
       bool doConstructorCompletion();
 
-      QList<DeclarationDepthPair> namespaceItems(KDevelop::DUContext* duContext, const KDevelop::CursorInRevision& position, bool global) const;
+      QList<DeclarationDepthPair> namespaceItems(KDevelop::DUContext* duContext, const KDevelop::CursorInRevision& position, bool global,
+                                                 const QSet<KDevelop::DUContext*>& skipContexts = QSet<KDevelop::DUContext*>()) const;
 
       AccessType m_accessType;
       QString m_expression;
diff --git a/languages/cpp/kdevcppsupport.desktop.cmake b/languages/cpp/kdevcppsupport.desktop.cmake
index e02c762..f3aae87 100644
--- a/languages/cpp/kdevcppsupport.desktop.cmake
+++ b/languages/cpp/kdevcppsupport.desktop.cmake
@@ -1,5 +1,6 @@
 [Desktop Entry]
 Type=Service
+Icon=text-x-c++src
 Exec=blubb
 Comment=C/C++ Language Support
 Comment[be]=Падтрымка мовы C/C++
@@ -117,7 +118,6 @@ GenericName[x-test]=xxLanguage Supportxx
 GenericName[zh_CN]=语言支持
 GenericName[zh_TW]=語言支援
 ServiceTypes=KDevelop/Plugin
-Icon=source-cpp
 X-KDE-Library=kdevcpplanguagesupport
 X-KDE-PluginInfo-Name=kdevcppsupport
 X-KDE-PluginInfo-Category=Language Support
diff --git a/languages/cpp/parser/parser.cpp b/languages/cpp/parser/parser.cpp
index c59bd78..abb720e 100644
--- a/languages/cpp/parser/parser.cpp
+++ b/languages/cpp/parser/parser.cpp
@@ -2172,10 +2172,8 @@ bool Parser::parseParameterDeclaration(ParameterDeclarationAST *&node)
   if (session->token_stream->lookAhead() == '=')
     {
       advance();
-      if (!parseLogicalOrExpression(expr,true))
-        {
-          //reportError(("Expression expected"));
-        }
+      //FIXME: this should actually be an InitializerClause as per http://www.nongnu.org/hcb/#parameter-declaration
+      parseLogicalOrExpression(expr, true) || parseBracedInitList(expr);
     }
 
   if( session->token_stream->lookAhead() != ',' && session->token_stream->lookAhead() != ')' && session->token_stream->lookAhead() != '>' )
diff --git a/languages/cpp/parser/rpp/pp-engine.cpp b/languages/cpp/parser/rpp/pp-engine.cpp
index 2b8ed16..e3eddcd 100644
--- a/languages/cpp/parser/rpp/pp-engine.cpp
+++ b/languages/cpp/parser/rpp/pp-engine.cpp
@@ -585,7 +585,7 @@ Value pp::eval_multiplicative(Stream& input)
       result *= value;
 
     } else if (token == '/') {
-      if (value.is_zero()) {
+      if (value.is_zero() || (!value.is_ulong() && !result.is_ulong() && value.l == -1 && result.l == LONG_LONG_MIN)) {
         KDevelop::ProblemPointer problem(new KDevelop::Problem);
         problem->setFinalLocation(KDevelop::DocumentRange(currentFileName(), KDevelop::SimpleRange(input.originalInputPosition().castToSimpleCursor(), 0)));
         problem->setDescription(i18n("Division by zero"));
@@ -598,7 +598,7 @@ Value pp::eval_multiplicative(Stream& input)
       }
 
     } else {
-      if (value.is_zero()) {
+      if (value.is_zero() || (!value.is_ulong() && !result.is_ulong() && value.l == -1 && result.l == LONG_LONG_MIN)) {
         KDevelop::ProblemPointer problem(new KDevelop::Problem);
         problem->setFinalLocation(KDevelop::DocumentRange(currentFileName(), KDevelop::SimpleRange(input.originalInputPosition().castToSimpleCursor(), 0)));
         problem->setDescription(i18n("Division by zero"));
diff --git a/languages/cpp/parser/tests/test_parser_cpp2011.cpp b/languages/cpp/parser/tests/test_parser_cpp2011.cpp
index b03a3d8..0be5b6f 100644
--- a/languages/cpp/parser/tests/test_parser_cpp2011.cpp
+++ b/languages/cpp/parser/tests/test_parser_cpp2011.cpp
@@ -414,6 +414,7 @@ void TestParser::testInitList_data()
   QTest::newRow("condition") << "void f() { if (std::complex<double> foo{1, 2}) {} }";
   QTest::newRow("temporary") << "auto s = std::string{};";
   QTest::newRow("temporary-template") << "auto c = std::complex<double>{1, 2};";
+  QTest::newRow("function-default-arg") << "void foo(bla arg = {}) {}";
 }
 
 void TestParser::testInitList()
diff --git a/languages/cpp/tests/test_cppcodecompletion.cpp b/languages/cpp/tests/test_cppcodecompletion.cpp
index 81be94b..56fe2e7 100644
--- a/languages/cpp/tests/test_cppcodecompletion.cpp
+++ b/languages/cpp/tests/test_cppcodecompletion.cpp
@@ -1179,6 +1179,28 @@ void TestCppCodeCompletion::testOverride()
   release(context);
 }
 
+void TestCppCodeCompletion::testOverrideDeleted()
+{
+  const QByteArray tmpl("class A {public: virtual void foo() = delete; };\n"
+                       "class B : public A { };");
+  TopDUContext* context = parse(tmpl, DumpNone);
+  DUChainWriteLocker lock;
+  DUContext* BCtx = context->childContexts().last();
+  QCOMPARE(BCtx->localScopeIdentifier().toString(), QLatin1String("B"));
+  CompletionItemTester tester(BCtx, "");
+  bool found = false;
+  foreach(CompletionItemTester::Item item, tester.items) {
+    Cpp::ImplementationHelperItem* override = dynamic_cast<Cpp::ImplementationHelperItem*>(item.data());
+    if (!override) {
+      continue;
+    }
+    found = true;
+    break;
+  }
+  QVERIFY(!found);
+  release(context);
+}
+
 void TestCppCodeCompletion::testSignalSlotCompletion() {
     // By processing qobjectdefs.h, we make sure that the qt-specific macros are defined in the duchain through overriding (see setuphelpers.cpp)
     addInclude("/qobjectdefs.h", "#define signals\n#define slots\n#define Q_SIGNALS\n#define Q_SLOTS\n#define Q_PRIVATE_SLOT\n#define SIGNAL\n#define SLOT\n int n;\n");
@@ -3794,4 +3816,20 @@ void TestCppCodeCompletion::testNestedInlineNamespace()
   release(top);
 }
 
+void TestCppCodeCompletion::testDuplicatedNamespace()
+{
+  // see also: https://bugs.kde.org/show_bug.cgi?id=328803
+  QByteArray test( "namespace foo { const int bar = 1; }\n"
+                   "namespace foo { const int asdf = 1; }\n"
+                   "namespace foo { }\n"
+                   "int main() {}\n" );
+  TopDUContext* top = parse(test, DumpNone);
+  DUChainWriteLocker lock;
+  CompletionItemTester tester(top->childContexts().last(), "foo::");
+  //TODO: the sort-order is apparently undefined...
+  qSort(tester.names);
+  QCOMPARE(tester.names, QStringList() << "asdf" << "bar" );
+  release(top);
+}
+
 #include "test_cppcodecompletion.moc"
diff --git a/languages/cpp/tests/test_cppcodecompletion.h b/languages/cpp/tests/test_cppcodecompletion.h
index 444516b..59807fd 100644
--- a/languages/cpp/tests/test_cppcodecompletion.h
+++ b/languages/cpp/tests/test_cppcodecompletion.h
@@ -149,6 +149,7 @@ private slots:
   void testParentConstructor();
   void testOverride_data();
   void testOverride();
+  void testOverrideDeleted();
   void testExecuteKeepWord_data();
   void testExecuteKeepWord();
   void testAfterVisibility_data();
@@ -158,6 +159,7 @@ private slots:
   void testLookaheadMatches();
   void testMemberAccessInstance();
   void testNestedInlineNamespace();
+  void testDuplicatedNamespace();
 public:
   enum DumpArea {
     DumpNone = 0,
diff --git a/projectbuilders/ninjabuilder/kcm_kdev_ninjabuilder.desktop b/projectbuilders/ninjabuilder/kcm_kdev_ninjabuilder.desktop
index 698d4ef..072c9c2 100644
--- a/projectbuilders/ninjabuilder/kcm_kdev_ninjabuilder.desktop
+++ b/projectbuilders/ninjabuilder/kcm_kdev_ninjabuilder.desktop
@@ -16,6 +16,7 @@ Name[ca]=Ninja
 Name[ca@valencia]=Ninja
 Name[da]=Ninja
 Name[de]=Ninja
+Name[el]=Nίντζα
 Name[es]=Ninja
 Name[fi]=Ninja
 Name[fr]=Ninja
@@ -23,7 +24,9 @@ Name[gl]=Ninja
 Name[hu]=Ninja
 Name[it]=Ninja
 Name[kk]=Ninja
+Name[nds]=Ninja
 Name[nl]=Ninja
+Name[pl]=Ninja
 Name[pt]=Ninja
 Name[pt_BR]=Ninja
 Name[sk]=Ninja
@@ -46,7 +49,9 @@ Comment[gl]=Configura os parámetros de Ninja
 Comment[hu]=Ninja beállítások módosítása
 Comment[it]=Configura le impostazioni di Ninja
 Comment[kk]=Ninja параметрлерін баптау
+Comment[nds]=Ninja-Instellen fastleggen
 Comment[nl]=Ninja-instellingen configureren
+Comment[pl]=Konfiguruj ustawienia Ninja
 Comment[pt]=Configurar as definições do Ninja
 Comment[pt_BR]=Configure as opções do Ninja
 Comment[sk]=Konfigurovať nastavenia Ninja
diff --git a/projectbuilders/ninjabuilder/kdevninja.desktop.cmake b/projectbuilders/ninjabuilder/kdevninja.desktop.cmake
index abb58c7..dac9e78 100644
--- a/projectbuilders/ninjabuilder/kdevninja.desktop.cmake
+++ b/projectbuilders/ninjabuilder/kdevninja.desktop.cmake
@@ -7,6 +7,7 @@ Name[ca@valencia]=Constructor de projectes Ninja
 Name[cs]=Překladač projektů Ninja
 Name[da]=Ninja-projektbygning
 Name[de]=Ninja-Projekterstellung
+Name[el]=Κατασκευαστής έργου Νίντζα
 Name[es]=Constructor de proyectos Ninja
 Name[fi]=Ninja-projektikäännin
 Name[fr]=Constructeur de projet Ninja
@@ -14,7 +15,9 @@ Name[gl]=Construtor de proxectos Ninja
 Name[hu]=Ninja projektfordító
 Name[it]=Compilatore progetto Ninja
 Name[kk]=Ninja жоба құрастырғышы
+Name[nds]=Ninja-Projektbuumoduul
 Name[nl]=Ninja-project bouwprogramma
+Name[pl]=Budowanie projektu Ninja
 Name[pt]=Construtor de Projectos Ninja
 Name[pt_BR]=Compilador de projetos Ninja
 Name[sk]=Zostavovač projektov Ninja
@@ -65,6 +68,7 @@ Comment[ca]=Constructor Ninja del KDevelop
 Comment[ca@valencia]=Constructor Ninja del KDevelop
 Comment[da]=Ninja-bygning til KDevelop
 Comment[de]=KDevelop-Ninja-Erstellung
+Comment[el]=Κατασκευαστής Νίντζα του KDevelop
 Comment[es]=Constructor Ninja para KDevelop
 Comment[fi]=KDevelop Ninja-käännin
 Comment[fr]=Constructeur Ninja pour KDevelop
@@ -72,7 +76,9 @@ Comment[gl]=Construtor Ninja para o KDevelop
 Comment[hu]=KDevelop Ninja fordító
 Comment[it]=Compilatore Ninja di KDevelop
 Comment[kk]=KDevelop-тың Ninja құрастырғышы
+Comment[nds]=Ninja-Buuwarktüüch vun KDevelop
 Comment[nl]=KDevelop Ninja-bouwprogramma
+Comment[pl]=Budowniczy Debugger dla KDevelop
 Comment[pt]=Construtor Ninja do KDevelop
 Comment[pt_BR]=Compilador Ninja do KDevelop
 Comment[sk]=Zostavovač KDevelop Ninja
diff --git a/projectbuilders/ninjabuilder/kdevninjabuilderplugin.cpp b/projectbuilders/ninjabuilder/kdevninjabuilderplugin.cpp
index 909d146..dc28408 100644
--- a/projectbuilders/ninjabuilder/kdevninjabuilderplugin.cpp
+++ b/projectbuilders/ninjabuilder/kdevninjabuilderplugin.cpp
@@ -20,7 +20,6 @@
 #include "ninjajob.h"
 #include <KAboutData>
 #include <KPluginFactory>
-#include <KStandardDirs>
 #include <KDebug>
 #include <KConfigGroup>
 #include <KShell>
@@ -43,7 +42,7 @@ KDevNinjaBuilderPlugin::KDevNinjaBuilderPlugin(QObject* parent, const QVariantLi
 
 bool KDevNinjaBuilderPlugin::hasError() const
 {
-    return KStandardDirs::findExe("ninja").isEmpty();
+    return NinjaJob::ninjaBinary().isEmpty();
 }
 
 static QStringList targetsInFolder(KDevelop::ProjectFolderItem* item)
diff --git a/projectbuilders/ninjabuilder/ninjajob.cpp b/projectbuilders/ninjabuilder/ninjajob.cpp
index 9efa57c..acb5d07 100644
--- a/projectbuilders/ninjabuilder/ninjajob.cpp
+++ b/projectbuilders/ninjabuilder/ninjajob.cpp
@@ -22,6 +22,7 @@
 #include <KDebug>
 #include <KLocalizedString>
 #include <KConfigGroup>
+#include <KStandardDirs>
 #include <interfaces/iproject.h>
 #include <outputview/outputmodel.h>
 #include <project/interfaces/ibuildsystemmanager.h>
@@ -44,7 +45,7 @@ NinjaJob::NinjaJob(KDevelop::ProjectBaseItem* item, const QStringList& arguments
     setFilteringStrategy( KDevelop::OutputModel::CompilerFilter );
     setProperties( NeedWorkingDirectory | PortableMessages | DisplayStderr | IsBuilderHint | PostProcessOutput );
 
-    *this << "ninja";
+    *this << ninjaBinary();
     *this << arguments;
 
     QStringList targets;
@@ -68,6 +69,15 @@ void NinjaJob::setIsInstalling( bool isInstalling )
     m_isInstalling = isInstalling;
 }
 
+QString NinjaJob::ninjaBinary()
+{
+    QString path = KStandardDirs::findExe("ninja-build");
+    if (path.isEmpty()) {
+        path = KStandardDirs::findExe("ninja");
+    }
+    return path;
+}
+
 KUrl NinjaJob::workingDirectory() const
 {
     KDevelop::ProjectBaseItem* it = item();
diff --git a/projectbuilders/ninjabuilder/ninjajob.h b/projectbuilders/ninjabuilder/ninjajob.h
index 0bd1725..1186c06 100644
--- a/projectbuilders/ninjabuilder/ninjajob.h
+++ b/projectbuilders/ninjabuilder/ninjajob.h
@@ -39,6 +39,7 @@ class NinjaJob : public KDevelop::OutputExecuteJob
     public:
         NinjaJob( KDevelop::ProjectBaseItem* item, const QStringList& arguments, const QByteArray& signal, QObject* parent );
         void setIsInstalling( bool isInstalling );
+        static QString ninjaBinary();
 
         virtual KUrl workingDirectory() const;
         virtual QStringList privilegedExecutionCommand() const;
diff --git a/projectmanagers/cmake/cmakebuilddirchooser.cpp b/projectmanagers/cmake/cmakebuilddirchooser.cpp
index 2ec9855..22b5de9 100644
--- a/projectmanagers/cmake/cmakebuilddirchooser.cpp
+++ b/projectmanagers/cmake/cmakebuilddirchooser.cpp
@@ -33,6 +33,7 @@ static const int maxExtraArgumentsInHistory = 15;
 CMakeBuildDirChooser::CMakeBuildDirChooser(QWidget* parent)
     : KDialog(parent)
 {
+    setDefaultButton(KDialog::Ok);
     setCaption(i18n("Configure a build directory"));
     
 //     QWidget* w= new QWidget(this);
@@ -95,7 +96,8 @@ QString CMakeBuildDirChooser::buildDirProject(const KUrl& srcDir)
     const QString pLine="CMAKE_HOME_DIRECTORY:INTERNAL=";
     while (!correct && !file.atEnd())
     {
-        QString line = file.readLine().trimmed();
+        // note: CMakeCache.txt is UTF8-encoded, also see bug 329305
+        QString line = QString::fromUtf8(file.readLine().trimmed());
         if(line.startsWith(pLine))
         {
             correct=true;
diff --git a/projectmanagers/cmake/cmakecommitchangesjob.cpp b/projectmanagers/cmake/cmakecommitchangesjob.cpp
index 09e2faa..a9c1898 100644
--- a/projectmanagers/cmake/cmakecommitchangesjob.cpp
+++ b/projectmanagers/cmake/cmakecommitchangesjob.cpp
@@ -170,6 +170,10 @@ KUrl::List CMakeCommitChangesJob::addProjectData(const CMakeProjectData& data)
         if(targetProps["FOLDER"]==QStringList("CTestDashboardTargets"))
             continue; //filter some annoying targets
 
+        if (!m_manager->filterManager()->isValid(KUrl(m_url, t.name), false, m_project)) {
+            continue;
+        }
+
         ProcessedTarget target;
         target.target = t;
         target.defines = targetProps["COMPILE_DEFINITIONS"];
diff --git a/projectmanagers/cmake/cmakeedit.cpp b/projectmanagers/cmake/cmakeedit.cpp
index 05cedc2..ff2d127 100644
--- a/projectmanagers/cmake/cmakeedit.cpp
+++ b/projectmanagers/cmake/cmakeedit.cpp
@@ -188,7 +188,9 @@ bool itemAffected(const ProjectBaseItem *item, const KUrl &changeUrl)
     if (listsPath.isEmpty())
         return false;
     
-    KUrl listsFolder(listsPath.toLocalFile(KUrl::AddTrailingSlash));
+    KUrl listsFolder(listsPath);
+    listsFolder = listsFolder.upUrl();
+
     //Who thought it was a good idea to have KUrl::isParentOf return true if the urls are equal?
     return listsFolder.QUrl::isParentOf(changeUrl);
 }
@@ -287,7 +289,7 @@ bool changesWidgetRemoveFileFromTarget(const ProjectBaseItem *item, ApplyChanges
     return followUses(widget->document(), targetRange, relative, lists, false, QString());
 }
 
-bool changesWidgetRemoveItems(const QList<ProjectBaseItem*> &items, ApplyChangesWidget *widget)
+bool changesWidgetRemoveItems(const QSet<ProjectBaseItem*> &items, ApplyChangesWidget *widget)
 {
     foreach(ProjectBaseItem *item, items)
     {
diff --git a/projectmanagers/cmake/cmakeedit.h b/projectmanagers/cmake/cmakeedit.h
index 23bf48a..9f35b22 100644
--- a/projectmanagers/cmake/cmakeedit.h
+++ b/projectmanagers/cmake/cmakeedit.h
@@ -39,7 +39,7 @@ namespace CMakeEdit {
     QList<KDevelop::ProjectBaseItem*> cmakeListedItemsAffectedByItemsChanged(const QList<KDevelop::ProjectBaseItem*> &items);
 
     bool changesWidgetRemoveFilesFromTargets(const QList<KDevelop::ProjectFileItem*> &files, KDevelop::ApplyChangesWidget *widget);
-    bool changesWidgetRemoveItems(const QList<KDevelop::ProjectBaseItem*> &items, KDevelop::ApplyChangesWidget *widget);
+    bool changesWidgetRemoveItems(const QSet<KDevelop::ProjectBaseItem*>& items, KDevelop::ApplyChangesWidget* widget);
     bool changesWidgetAddFolder(const KUrl &folderUrl, const CMakeFolderItem *toFolder, KDevelop::ApplyChangesWidget *widget);
     bool changesWidgetRemoveCMakeFolder(const CMakeFolderItem *folder, KDevelop::ApplyChangesWidget *widget);
     bool changesWidgetAddFolder(const KUrl &folderUrl, const CMakeFolderItem *toFolder, KDevelop::ApplyChangesWidget *widget);
@@ -50,4 +50,4 @@ namespace CMakeEdit {
     KUrl afterMoveUrl(const KUrl& origUrl, const KUrl& movedOrigUrl, const KUrl& movedNewUrl);
 }
 
-#endif
\ No newline at end of file
+#endif
diff --git a/projectmanagers/cmake/cmakeimportjob.cpp b/projectmanagers/cmake/cmakeimportjob.cpp
index f5c5b7d..ae621d2 100644
--- a/projectmanagers/cmake/cmakeimportjob.cpp
+++ b/projectmanagers/cmake/cmakeimportjob.cpp
@@ -109,12 +109,14 @@ void CMakeImportJob::importFinished()
 void CMakeImportJob::initialize()
 {
     ReferencedTopDUContext ctx;
-    if(m_dom->url() == m_project->folder()) {
-        ctx = initializeProject(dynamic_cast<CMakeFolderItem*>(m_dom));
-    } else {
+    ProjectBaseItem* parent = m_dom->parent();
+    while (parent && !ctx) {
         DUChainReadLocker lock;
-        ctx = DUChain::self()->chainForDocument(KUrl(m_dom->parent()->url(), "CMakeLists.txt"));
-        Q_ASSERT(ctx);
+        ctx = DUChain::self()->chainForDocument(KUrl(parent->url(), "CMakeLists.txt"));
+        parent = parent->parent();
+    }
+    if (!ctx) {
+        ctx = initializeProject(dynamic_cast<CMakeFolderItem*>(m_dom));
     }
     importDirectory(m_project, m_dom->url(), ctx);
 }
@@ -211,7 +213,6 @@ CMakeCommitChangesJob* CMakeImportJob::importDirectory(IProject* project, const
 {
     Q_ASSERT(thread() == m_project->thread());
     KUrl cmakeListsPath(url, "CMakeLists.txt");
-    
     CMakeCommitChangesJob* commitJob = new CMakeCommitChangesJob(url, m_manager, project);
     commitJob->moveToThread(thread());
     m_jobs += commitJob;
@@ -227,6 +228,11 @@ CMakeCommitChangesJob* CMakeImportJob::importDirectory(IProject* project, const
             if (!m_manager->filterManager()->isValid(folder, true, project)) {
                 continue;
             }
+            KUrl newcmakeListsPath(folder, "CMakeLists.txt");
+            if(!QFile::exists(newcmakeListsPath.toLocalFile())) {
+               kWarning() << "Unable to open " << newcmakeListsPath.toLocalFile();
+               continue;
+            }
             CMakeCommitChangesJob* job = importDirectory(project, folder, ctx);
             job->setFindParentItem(false);
             connect(commitJob, SIGNAL(folderCreated(KDevelop::ProjectFolderItem*)),
diff --git a/projectmanagers/cmake/cmakemanager.cpp b/projectmanagers/cmake/cmakemanager.cpp
index 6880364..6773a44 100644
--- a/projectmanagers/cmake/cmakemanager.cpp
+++ b/projectmanagers/cmake/cmakemanager.cpp
@@ -201,6 +201,7 @@ KDevelop::ProjectFolderItem* CMakeManager::import( KDevelop::IProject *project )
     connect(w, SIGNAL(fileChanged(QString)), SLOT(dirtyFile(QString)));
     connect(w, SIGNAL(directoryChanged(QString)), SLOT(directoryChanged(QString)));
     m_watchers[project] = w;
+    kDebug(9042) << "Added watcher for project " << project << project->name();
     m_filter->add(project);
     
     KUrl cachefile=CMake::currentBuildDir(project);
@@ -588,7 +589,6 @@ bool CMakeManager::removeFilesAndFolders(const QList<KDevelop::ProjectBaseItem*>
     foreach(ProjectBaseItem* item, items)
     {
         Q_ASSERT(item->folder() || item->file());
-        Q_ASSERT(!item->file() || !item->file()->parent()->target());
 
         urls += item->url();
         if(!p)
@@ -600,7 +600,7 @@ bool CMakeManager::removeFilesAndFolders(const QList<KDevelop::ProjectBaseItem*>
     changesWidget.setCaption(DIALOG_CAPTION);
     changesWidget.setInformation(i18n("Remove files and folders from CMakeLists as follows:"));
 
-    bool cmakeSuccessful = changesWidgetRemoveItems(cmakeListedItemsAffectedByItemsChanged(items), &changesWidget);
+    bool cmakeSuccessful = changesWidgetRemoveItems(cmakeListedItemsAffectedByItemsChanged(items).toSet(), &changesWidget);
 
     if (changesWidget.hasDocuments() && cmakeSuccessful)
         cmakeSuccessful &= changesWidget.exec() && changesWidget.applyAllChanges();
@@ -845,6 +845,8 @@ void CMakeManager::projectClosing(IProject* p)
     delete m_watchers.take(p);
 
     m_filter->remove(p);
+
+    kDebug(9042) << "Project closed" << p;
 }
 
 QStringList CMakeManager::processGeneratorExpression(const QStringList& expr, IProject* project, ProjectTargetItem* target) const
@@ -875,7 +877,12 @@ CMakeFolderItem* CMakeManager::takePending(const KUrl& url)
 
 void CMakeManager::addWatcher(IProject* p, const QString& path)
 {
-    m_watchers[p]->addPath(path);
+    if (QFileSystemWatcher* watcher = m_watchers.value(p)) {
+        watcher->addPath(path);
+    } else {
+        kWarning() << "Could not find a watcher for project" << p << p->name() << ", path " << path;
+        Q_ASSERT(false);
+    }
 }
 
 CMakeProjectData CMakeManager::projectData(IProject* project)
diff --git a/projectmanagers/cmake/kcm_kdevcmake_settings.desktop b/projectmanagers/cmake/kcm_kdevcmake_settings.desktop
index c22fb80..a71c382 100644
--- a/projectmanagers/cmake/kcm_kdevcmake_settings.desktop
+++ b/projectmanagers/cmake/kcm_kdevcmake_settings.desktop
@@ -84,7 +84,7 @@ Comment[nb]=Sett opp CMake-bygger
 Comment[nds]=CMake instellen
 Comment[ne]=सीमेक सेटिङ कन्फिगर गर्नुहोस्
 Comment[nl]=CMake-instellingen configureren
-Comment[pl]=Ustawienia CMake
+Comment[pl]=Konfiguruj ustawienia CMake
 Comment[pt]=Configurar as opções do CMake
 Comment[pt_BR]=Configurar as opções do CMake
 Comment[ru]=Настройка параметров CMake
diff --git a/projectmanagers/cmake/kdevcmakemanager.desktop.cmake b/projectmanagers/cmake/kdevcmakemanager.desktop.cmake
index 5a5f7b9..c5b7690 100644
--- a/projectmanagers/cmake/kdevcmakemanager.desktop.cmake
+++ b/projectmanagers/cmake/kdevcmakemanager.desktop.cmake
@@ -22,7 +22,7 @@ Name[km]=កម្មវិធី​គម្រោង CMake
 Name[nb]=CMake prosjekstyrer
 Name[nds]=CMake-Projektpleger
 Name[nl]=CMake-projectbeheerder
-Name[pl]=Menadżer projektów CMake
+Name[pl]=Zarządca projektów CMake
 Name[pt]=Gestor de Projectos do CMake
 Name[pt_BR]=Gerenciador de Projeto CMake
 Name[ru]=Управление проектами CMake
@@ -60,7 +60,7 @@ GenericName[nb]=Prosjektbehandler
 GenericName[nds]=Projektpleger
 GenericName[nl]=Projectbeheerder
 GenericName[nn]=Prosjekthandsamar
-GenericName[pl]=Menadżer projektu
+GenericName[pl]=Zarządca projektu
 GenericName[pt]=Gestor de Projectos
 GenericName[pt_BR]=Gerenciador de Projetos
 GenericName[ru]=Управление проектами
diff --git a/projectmanagers/custommake/custommakemanager.cpp b/projectmanagers/custommake/custommakemanager.cpp
index fb1cd91..f697fc0 100644
--- a/projectmanagers/custommake/custommakemanager.cpp
+++ b/projectmanagers/custommake/custommakemanager.cpp
@@ -62,6 +62,9 @@ CustomMakeManager::CustomMakeManager( QObject *parent, const QVariantList& args
     Q_ASSERT(i);
     d->m_builder = i->extension<IMakeBuilder>();
     Q_ASSERT(d->m_builder);
+
+    connect(this, SIGNAL(reloadedFileItem(KDevelop::ProjectFileItem*)),
+            this, SLOT(reloadMakefile(KDevelop::ProjectFileItem*)));
 }
 
 CustomMakeManager::~CustomMakeManager()
@@ -131,25 +134,52 @@ QList<ProjectTargetItem*> CustomMakeManager::targets(KDevelop::ProjectFolderItem
     return ret;
 }
 
-ProjectFileItem* CustomMakeManager::createFileItem(IProject* project, const KUrl& url, ProjectBaseItem* parent)
+static bool isMakefile(const QString& fileName)
 {
-    KDevelop::ProjectFileItem *item = new KDevelop::ProjectFileItem( project, url, parent );
-    const QString fileName = url.fileName();
-    if( fileName == QLatin1String("Makefile")
+    return  ( fileName == QLatin1String("Makefile")
         || fileName == QLatin1String("makefile")
         || fileName == QLatin1String("GNUmakefile")
-        || fileName == QLatin1String("BSDmakefile") )
+        || fileName == QLatin1String("BSDmakefile") );
+}
+
+void CustomMakeManager::createTargetItems(IProject* project, const KUrl& url, ProjectBaseItem* parent)
+{
+    Q_ASSERT(isMakefile(url.fileName()));
+    foreach(const QString& target, parseCustomMakeFile( url ))
     {
-        QStringList targetlist = parseCustomMakeFile( url );
-        foreach( const QString &target, targetlist )
-        {
-            new CustomMakeTargetItem( project, target, parent );
-//             d->m_testItems.append( targetItem ); // debug
+        if(!isValid(KUrl(parent->url(), target), false, project)){
+            continue;
         }
+        new CustomMakeTargetItem( project, target, parent );
+//         d->m_testItems.append( targetItem ); // debug
+    }
+}
+
+ProjectFileItem* CustomMakeManager::createFileItem(IProject* project, const KUrl& url, ProjectBaseItem* parent)
+{
+    ProjectFileItem* item = new ProjectFileItem(project, url, parent);
+    if (isMakefile(url.fileName())){
+        createTargetItems(project, url, parent);
     }
     return item;
 }
 
+void CustomMakeManager::reloadMakefile(ProjectFileItem* file)
+{
+    if( !isMakefile(file->url().fileName())){
+        return;
+    }
+    ProjectBaseItem* parent = file->parent();
+    // remove the items that are Makefile targets
+    foreach(ProjectBaseItem* item, parent->children()){
+        if (item->target()){
+            delete item;
+        }
+    }
+    // Recreate the targets.
+    createTargetItems(parent->project(), file->url(), parent);
+}
+
 ProjectFolderItem* CustomMakeManager::createFolderItem(IProject* project, const KUrl& url, ProjectBaseItem* parent)
 {
     // TODO more faster algorithm. should determine whether this directory
diff --git a/projectmanagers/custommake/custommakemanager.h b/projectmanagers/custommake/custommakemanager.h
index a63aca8..48d673c 100644
--- a/projectmanagers/custommake/custommakemanager.h
+++ b/projectmanagers/custommake/custommakemanager.h
@@ -115,13 +115,16 @@ protected:
     virtual KDevelop::ProjectFileItem* createFileItem(KDevelop::IProject* project, const KUrl& url, KDevelop::ProjectBaseItem* parent);
     virtual KDevelop::ProjectFolderItem* createFolderItem(KDevelop::IProject* project, const KUrl& url, KDevelop::ProjectBaseItem* parent = 0);
 
+private slots:
+    void reloadMakefile(KDevelop::ProjectFileItem *item);
+
 private:
     /**
      * Initialize targets by reading Makefile in @arg dir
      * @return Target lists in Makefile at @arg dir.
      */
     QStringList parseCustomMakeFile( const KUrl &makefile );
-
+    void createTargetItems(KDevelop::IProject* project, const KUrl& url, KDevelop::ProjectBaseItem* parent);
 private:
     class Private;
     Private *d;
diff --git a/projectmanagers/custommake/kdevcustommakemanager.desktop.cmake b/projectmanagers/custommake/kdevcustommakemanager.desktop.cmake
index 1b7308b..6278e2a 100644
--- a/projectmanagers/custommake/kdevcustommakemanager.desktop.cmake
+++ b/projectmanagers/custommake/kdevcustommakemanager.desktop.cmake
@@ -58,7 +58,7 @@ GenericName[nb]=Prosjektbehandler
 GenericName[nds]=Projektpleger
 GenericName[nl]=Projectbeheerder
 GenericName[nn]=Prosjekthandsamar
-GenericName[pl]=Menadżer projektu
+GenericName[pl]=Zarządca projektu
 GenericName[pt]=Gestor de Projectos
 GenericName[pt_BR]=Gerenciador de Projetos
 GenericName[ru]=Управление проектами
diff --git a/providers/ghprovider/kdevghprovider.desktop.cmake b/providers/ghprovider/kdevghprovider.desktop.cmake
index 2b8667e..9fe8336 100644
--- a/providers/ghprovider/kdevghprovider.desktop.cmake
+++ b/providers/ghprovider/kdevghprovider.desktop.cmake
@@ -8,6 +8,7 @@ Comment[ca]=Aquest connector ajuda a obtenir els projectes des de Github
 Comment[ca@valencia]=Este connector ajuda a obtindre els projectes des de Github
 Comment[da]=Dette plugin hjælper med at hente projekter fra Github
 Comment[de]=Dieses Modul hilft dabei Github-Projekte zu beziehen
+Comment[el]=Αυτό το πρόσθετο βοηθάει στην απόκτηση έργων από το Github
 Comment[es]=Este complemento ayuda a obtener proyectos desde Github
 Comment[fi]=Tämä liitännäinen auttaa Github-projektien hakemisessa
 Comment[fr]=Ce module externe permet d'obtenir des projets depuis Github
@@ -17,6 +18,7 @@ Comment[it]=Questa estensione consente di ottenere progetti da Github
 Comment[kk]=Github жобаларын жүктеп алуға көмек плагині
 Comment[nb]=Dette programtillegget hjelper til med å skaffe prosjekter fra Github
 Comment[nl]=Deze plug-in helpt om projecten uit Github te verkrijgen
+Comment[pl]=Wtyczka ta pomaga przy pobieraniu projektów z Githuba
 Comment[pt]=Este 'plugin' ajuda a obter projectos do GitHub
 Comment[pt_BR]=Este plugin ajuda a obter projetos a partir do Github
 Comment[sk]=Tento modul pomáha získavať projekty z Github
@@ -32,6 +34,7 @@ Name[ca@valencia]=Proveïdor de Github
 Name[cs]=Poskytovatel Github
 Name[da]=Github-udbyder
 Name[de]=Github-Provider
+Name[el]=Πάροχος Github
 Name[es]=Proveedor de Github
 Name[fi]=Github-tarjoaja
 Name[fr]=Fournisseur Github
@@ -40,7 +43,9 @@ Name[hu]=Github szolgáltató
 Name[it]=Fornitore Github
 Name[kk]=Github таратушысы
 Name[nb]=Github-leverandør
+Name[nds]=Github-Anbeder
 Name[nl]=Github-leverancier
+Name[pl]=Dostawca Github
 Name[pt]=Fornecedor do GitHub
 Name[pt_BR]=Fornecedor do Github
 Name[sk]=Poskytovateľ Github
@@ -56,6 +61,7 @@ GenericName[ca@valencia]=Proveïdor de Github
 GenericName[cs]=Poskytovatel Github
 GenericName[da]=Github-udbyder
 GenericName[de]=Github-Provider
+GenericName[el]=Πάροχος Github
 GenericName[es]=Proveedor de Github
 GenericName[fi]=Github-tarjoaja
 GenericName[fr]=Fournisseur Github
@@ -64,7 +70,9 @@ GenericName[hu]=Github szolgáltató
 GenericName[it]=Fornitore Github
 GenericName[kk]=Github таратушысы
 GenericName[nb]=Github-leverandør
+GenericName[nds]=Github-Anbeder
 GenericName[nl]=Github-leverancier
+GenericName[pl]=Dostawca Github
 GenericName[pt]=Fornecedor do GitHub
 GenericName[pt_BR]=Fornecedor do Github
 GenericName[sk]=Poskytovateľ Github
